src/librustc_mir/borrow_check/diagnostics/mod.rs

```diff
@@ -834,15 +856,41 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                         && parent.map_or(false, |p| {
                             tcx.lang_items().group(LangItemGroup::Op).contains(&p)
                         });
+                    let is_deref =
+                        !from_hir_call && tcx.is_diagnostic_item(sym::deref_method, method_did);
                     let fn_call_span = *fn_span;
 
                     let self_arg = tcx.fn_arg_names(method_did)[0];
 
+                    debug!(
+                        "terminator = {:?} from_hir_call={:?}",
+                        self.body[location.block].terminator, from_hir_call
+                    );
+
                     let kind = if is_fn_once {
-                        FnSelfUseKind::FnOnceCall
+                        Some(FnSelfUseKind::FnOnceCall)
                     } else if is_operator {
-                        FnSelfUseKind::Operator { self_arg }
+                        Some(FnSelfUseKind::Operator { self_arg })
+                    } else if is_deref {
+                        let deref_target =
+                            tcx.get_diagnostic_item(sym::deref_target).and_then(|deref_target| {
+                                Instance::resolve(tcx, self.param_env, deref_target, method_substs)
+                                    .transpose()
+                            });
+                        if let Some(Ok(instance)) = deref_target {
+                            let deref_ty = instance.ty(tcx, self.param_env);
+                            Some(FnSelfUseKind::DerefCoercion {
+                                deref_target: tcx.def_span(instance.def_id()),
+                                deref_ty,
+                            })
+                        } else {
+                            None
+                        }
                     } else {
+                        None
+                    };
```

Comment:
Can we add a comment for this block? It's getting too long to quickly realize what's going on.